feat: AI model card

antoinemeyer5 opened the pull request:
## What does this PR do?

Features:
- Rename `TensorFlow` occurrences with `AiModel`
  - Still retro compatibility with tensorflowModel
- Create `AiModelCard` class and tools (access control implemented)
- Export `TensorboardLogs` data into CSV
- Test new `AiModelCard` code

## Related issues

[N/A](https://github.com/usnistgov/WIPP/issues/37)

Commit:
fix: update keycloak and springdoc urls

Commit:
feat: add model cards data

Commit:
refactor: change TensorflowModel into AIModel

Commit:
feat: add enum in AIModel to select machine learning library

Commit:
feat: add attributes to ModelCards

Commit:
fix: finish conversion from tensorflowmodel to aimodel

Commit:
fix: remove unused model cards controller and repository

Commit:
fix: change 'tensorflow' to 'ai' strings

Commit:
feat: init the logic to create model card after workflow is done

Commit:
feat: init the test model card

Commit:
feat: rename AIModel into AiModel

Commit:
fix: better name for AiModel parameter framework

Commit:
fix: better name and fix tests

Commit:
fix: better name and fix tests

Commit:
fix: better name and fix class

Commit:
feat: adding model card creation in the pipeline

Commit:
feat: setup model card format version 1

Commit:
feat: export for tensorflow platform

Commit:
feat: export tensorflow model card

Commit:
feat: export huggingface model card

Commit:
feat: export bioimage model card and add fields to tensorflow and huggingface

Commit:
feat: handling inputs and outputs fields

Commit:
feat: handling more fields

Commit:
feat: export tensorboardslogs into csv files

Commit:
feat: adding GET to create and get csv TensorBoardLogs data

Commit:
feat: rename create to export for consistency

Commit:
feat: Automatic export when TensorBoard finish his work; Add comments

Commit:
fix: tensboard url in config

Commit:
feat: fill model card with tensorboard logs informations

Commit:
fix: rename ModelCard in AiModelCard

Commit:
feat: fill in the ai model card

antoinemeyer5 committed:
feat: add fields to plugin manifest

Commit:
fix: rename `type` in `operationType`

Commit:
feat: use framework and architecture from plugin manifest

Commit:
feat: add logic for different export and tests

Commit:
feat: add tests for AiModelCard

Commit:
fix: better naming and license

Commit:
fix: add new license

Commit:
fix: license and naming

Commit:
fix: license and naming

antoinemeyer5 reviewed (commented):
LGTM

@MyleneSimon, I do not have the rights to add reviewers. Can you review this one?

Commit:
fix: better comments

Commit:
fix: fix

Commit:
feat: handle multi operationType and split training data and params

antoinemeyer5 commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,200 @@
+        // Fill with TensorboardLogs data
+        try {
+            // Declare id
+            String id = "6682f3d43149955bd95f59ab"; // todo: use job.getId()
```
replace

Commit:
feat: backward compatibility with existing plugins

Commit:
fix: splitting into functions

antoinemeyer5 commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+        if(outputs_options!=null && !outputs_options.isEmpty()) {
+            aiModel.setFramework(outputs_options.get("framework").toString());
+        } else { aiModel.setFramework("N/A"); }
+        aiModelRepository.save(aiModel);
```
This line should be in L91

antoinemeyer5 commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCardController.java:
```diff
@@ -0,0 +1,231 @@
+        if(!omc.isPresent()){
+            throw new ResourceNotFoundException("ModelCard not found.");
+        }
+        AiModelCard mc = omc.get();
```
Create function for this part (L78 to L83) and use it in each methods

MyleneSimon commented on wipp-backend-core/src/main/java/gov/nist/itl/ssd/wipp/backend/core/CoreConfig.java:
```diff
@@ -15,6 +15,7 @@ public class CoreConfig {
 
     public static final String BASE_URI = "/api";
     public static final String PYRAMIDS_BASE_URI = "/pyramid-files";
+    public static final String TENSORBOARD_URI = "http://localhost:6006";
```
This should be a configurable property instead of hardcoded value
So something like:
```
@Value("{$tensorboard.uri}")
private String tensorboardUri
```
and then add matching getter, and property in application.properties and pom files

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModel.java:
```diff
@@ -0,0 +1,124 @@
+	private Date creationDate;
+
+	@Indexed(unique = true, sparse = true)
+    @ManualRef(Job.class)
```
formatting

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+
+/**
+ * @author Mohamed Ouladi <mohamed.ouladi at nist.gov>
+ * @author Mylene Simon <mylene.simon at nist.gov>
```
you can add yourself to the list of authors here

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+	@Autowired
+	private AiModelRepository aiModelRepository;
+
+    @Autowired
```
formatting

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+        assert plugin != null;
+
+        AiModel aiModel = new AiModel(job, outputName);
+		// Set owner to job owner
```
formatting

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+        aiModel.setOwner(job.getOwner());
+        // Set TM to private
+        aiModel.setPubliclyShared(false);
+        // Set framework (WARNING!: first output framework used to set AiModel framework)
```
Can you get the correct output using the outputName instead? otherwise this will fail in some cases

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelDataHandler.java:
```diff
@@ -0,0 +1,220 @@
+        setFramework(plugin, aiModel);
+
+        File trainedModelFolder = new File(config.getAiModelsFolder(), aiModel.getId());
+		trainedModelFolder.mkdirs();
```
formatting

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelEventHandler.java:
```diff
@@ -0,0 +1,125 @@
+public class AiModelEventHandler {
+	
+	@Autowired
+    AiModelRepository aiModelRepository;
```
formatting

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodel/AiModelLogic.java:
```diff
@@ -0,0 +1,53 @@
+
+    public void assertAiModelNameUnique(String name) {
+        if (aiModelRepository.countByName(name) != 0) {
+            throw new ClientException("A ai model named \""
```
AI instead of ai

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCard.java:
```diff
@@ -0,0 +1,201 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
+    private Date date;
+
+    /* Platform where the model will be published. */
```
AI framework

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCard.java:
```diff
@@ -0,0 +1,201 @@
+        this.description = plugin.getDescription();
+        this.citation = plugin.getCitation();
+        this.operationType= plugin.getOperationType();
+        // WARNING!: first output architecture used to set AiModel architecture
```
To be fixed (cf previous comment)

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCardRepository.java:
```diff
@@ -0,0 +1,52 @@
+    void delete(AiModelCard mc);
+
+    @Query("{aiModelId:'?0'}")
+    AiModelCard findModelCardByAiModelId(String aiModelId);
```
I think that you should be able to drop the query for now if you name the method `findByAiModelId(@Param("aiModelId") String aiModelId`
Once you add access control you would need a custom query indeed

antoinemeyer5 replied:
I have access control now!

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCardRepository.java:
```diff
@@ -0,0 +1,52 @@
+ */
+@Tag(name="AiModelCard Entity")
+@RepositoryRestResource
+public interface AiModelCardRepository extends PrincipalFilteredRepository<AiModelCard, String> {
```
This extends PrincipalFilteredRepository but I did not see an `owner` field in the AiModelCard class? It could be added and handled the same way TensorboardLogs owner is handled (ie matching the ownership of AiModel)

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/tensorboard/TensorBoardLogsController.java:
```diff
@@ -0,0 +1,139 @@
+     * @param id
+     * @throws IOException
+     */
+    @RequestMapping(value = "export/csv", method = RequestMethod.GET)
```
a `@PreAuthorize` should be added to the two methods to enable access control (checkout for example the AiModel download controller)

MyleneSimon commented on wipp-backend-data/src/main/java/gov/nist/itl/ssd/wipp/backend/data/aimodelcard/AiModelCardController.java:
```diff
@@ -0,0 +1,231 @@
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_VALUE
+    )
+    public ResponseEntity<byte[]> tensorflow(@PathVariable("id") String id) throws IOException
```
`@PreAuthorize` should be added to both methods for access control

MyleneSimon reviewed (commented):
Overall looking good, a few minor comments.
There is no access control on the model cards in this version, which is fine for a demo but we will want to add it before merging in the main branch (I put a few pointers in the comments)

Commit:
fix: Mylene's reviews

Commit:
fix: test

Commit:
feat: add access control

antoinemeyer5 commented:
> Overall looking good, a few minor comments. There is no access control on the model cards in this version, which is fine for a demo but we will want to add it before merging in the main branch (I put a few pointers in the comments)

Thanks for your review.

- I have corrected your reviews.
- I have added the 'owner' and 'publiclyShared' fields to the AI model card Java class.
- I have added access control.

Let's go through it again and let me know if you notice anything before we merge it.

Commit:
feat: refactor code and export CDCS record

Commit:
fix: fix and remove todo typo

Commit:
feat: create AI model card through API

Commit:
fix: undo bad commit
